src/components/AnnotationViewer.vue

```diff
@@ -1047,6 +1070,433 @@ export default class AnnotationViewer extends Vue {
     });
   }
 
+  // Helper function to find connected components using Union-Find
+  private findConnectedComponents(
+    connections: IAnnotationConnection[],
+  ): { annotations: Set<string>; connections: IAnnotationConnection[] }[] {
+    // Simple Union-Find implementation
+    const parent = new Map<string, string>();
+
+    // Find with path compression
+    function find(x: string): string {
+      if (!parent.has(x)) {
+        parent.set(x, x);
+      }
+      if (parent.get(x) !== x) {
+        parent.set(x, find(parent.get(x)!));
+      }
+      return parent.get(x)!;
+    }
+
+    // Union operation
+    function union(x: string, y: string): void {
+      parent.set(find(x), find(y));
+    }
+
+    // Process all connections
+    connections.forEach((conn) => {
+      union(conn.parentId, conn.childId);
+    });
+
+    // Group annotations and their connections by root
+    const components = new Map<
+      string,
+      {
+        annotations: Set<string>;
+        connections: IAnnotationConnection[];
+      }
+    >();
+
+    // Initialize components with annotations
+    parent.forEach((_, node) => {
+      const root = find(node);
+      if (!components.has(root)) {
+        components.set(root, {
+          annotations: new Set(),
+          connections: [],
+        });
+      }
+      components.get(root)!.annotations.add(node);
+    });
+
+    // Add connections to their respective components
+    connections.forEach((conn) => {
+      const root = find(conn.parentId); // Could use either parent or child
+      components.get(root)!.connections.push(conn);
+    });
+
+    return Array.from(components.values());
+  }
+
+  // This helper function collects just the annotations that are displayed
+  // but across all time, which is important for timelapse mode.
+  getDisplayedAnnotationIdsAcrossTime(): Set<string> {
+    const totalAnnotationIdsSet: Set<string> = new Set();
+    for (const layer of this.validLayers) {
+      if (layer.visible || this.showAnnotationsFromHiddenLayers) {
+        for (const annotation of this.displayableAnnotations) {
+          if (annotation.channel === layer.channel) {
+            const sliceIndexes = this.store.layerSliceIndexes(layer);
+            if (
+              (this.store.unrollXY ||
+                annotation.location.XY === sliceIndexes?.xyIndex) &&
+              (this.store.unrollZ ||
+                annotation.location.Z === sliceIndexes?.zIndex)
+            ) {
+              totalAnnotationIdsSet.add(annotation.id);
```

Comment:
Can an annotation be on several layers ? If no, I wonder if it would not be more efficient to iterate over annotations instead of layers first, to be able to break your loop on layers once you found if your annotation belongs to a displayed layer  
Maybe to check if we see performances issues

Reply:
Yes, unfortunately, annotations can exist on several layers, which makes it a bit more complex. Indeed, that is often used for some kinds of timelapse analysis, where we make two layers that correspond to different times so we can see annotations from two times at once. But this probably can be optimized if needed. Shall I leave a comment about it?